### PR TITLE
Support for mTLS

### PR DESCRIPTION
**Pull request by nickhyoti**
Add optional support for mTLS

## What does this PR accomplish?
By adding mTLS, this mandates the use of a certificate to be provided by a reverse proxy (or any client connecting to Concourse).  This reduces the attack surface for Concourse and also reduces the risk of an incorrect reverse proxy being connected to Concourse

Feature was raised in [this RFC](https://github.com/concourse/rfcs/pull/73)

## Changes proposed by this PR:

Additional command line option added tls-ca-cert in atc/atccmd/command.go
When specified, this needs to be set to allow the certificate & private key to be validated on connection.

## Notes to reviewer:

To test, use [this example](https://github.com/nicholasjackson/mtls-go-example) to create the necessary certificates and keys.
Testing locally, the Common Name for the certificates can be set to localhost.

When creating the certificates, there will be three directories created

4_client - this will contain two files of interest certs/localhost.cert.pem and private/localhost.key.pem.  These will be used by the reverse proxy

3_application  - this will contain two files of interest certs/localhost.cert.pem and private/localhost.key.pem.  These can be used by Concourse web to enable TLS using parameters tls-cert and tls-key.

2_intermediate - this will contain certs/ca-chain.cert.pem.  This should be passed as the parameter tls-ca-cert.  Once enabled, any client will have to provide the key and certificates in the 4_client directory.

As an example of testing connections using curl with the URL set to localhost:8443 in docker-compose.yml
```
curl  -k https://localhost:8443/
curl: (56) OpenSSL SSL_read: error:14094412:SSL routines:ssl3_read_bytes:sslv3 alert bad certificate, errno 0
```
The connection has been rejected

When the correct parameters are provided
```
curl  --cacert 2_intermediate/certs/ca-chain.cert.pem --cert 4_client/certs/localhost.cert.pem --key 4_client/private/localhost.key.pem https://localhost:8443/
```
Here the connection is accepted.  The parameter cacert is included to allow self signed certificates to be used.

The testflight scripts have also been amended to allow a cacert parameter to be provided 

## Release Note
* Added optional support for mTLS to mandate use of certificate from reverse proxy to connect to Concourse

## Contributor Checklist

- [x] Followed [Code of conduct], [Contributing Guide] & avoided [Anti-patterns]
- [x] [Signed] all commits
- [x] Added tests (Unit and/or Integration)
- [x] Updated [Documentation]
- [x] Added release note (Optional)

[Code of Conduct]: https://github.com/concourse/concourse/blob/master/CODE_OF_CONDUCT.md
[Contributing Guide]: https://github.com/concourse/concourse/blob/master/CONTRIBUTING.md
[Anti-patterns]: https://github.com/concourse/concourse/wiki/Anti-Patterns
[Signed]: https://help.github.com/en/github/authenticating-to-github/signing-commits
[Documentation]: https://github.com/concourse/docs

## Reviewer Checklist

- [ ] Code reviewed
- [ ] Tests reviewed
- [ ] Documentation reviewed
- [ ] Release notes reviewed
- [ ] PR acceptance performed
- [ ] New config flags added? Ensure that they are added to the
  [BOSH](https://github.com/concourse/concourse-bosh-release) and
  [Helm](https://github.com/concourse/helm) packaging; otherwise, ignored for
  the [integration
  tests](https://github.com/concourse/ci/tree/master/tasks/scripts/check-distribution-env)
  (for example, if they are Garden configs that are not displayed in the
  `--help` text).

### CODE REVIEW

**Comment by nickhyoti** [comment]
tests now passing

**Comment by nickhyoti** [comment]
Please advise on the timetable for this issue.  Is there anything that is needed from our side?

**Comment by taylorsilva** [comment]
@nickhyoti Hey, the team is back from vacation/holidays now! December was a slow month. @muntac or someone else will get around to this. Thanks so much for your work on this!

**Comment by nickhyoti** [comment]
Any thoughts on this?  I get asked by my mgr about this

**Comment by taylorsilva** [comment]
Unassigning muntac. We'll re-assign at standup tomorrow.

File: testflight/suite_test.go [hunk 1]
```diff
@@ -104,6 +105,11 @@ var _ = SynchronizedBeforeSuite(func() []byte {
 		config.ATCGuestPassword = atcGuestPassword
 	}
 
+	caCert :=  os.Getenv("CA_CERT")
```
Comment (taylorsilva) [inline testflight/suite_test.go:108]: If we move forward with putting mutual tls as an override file then we don't need to update testflight to handle mutual tls, which I would prefer. If we want to test mutual tls we should add a topgun test. There are some changes happening to topgun right now so I wouldn't bother adding test coverage for that **for now.**

File: Dockerfile [hunk 2]
```diff
@@ -21,6 +21,10 @@ ENV CONCOURSE_TSA_WORKER_PRIVATE_KEY  /concourse-keys/worker_key
 
 # download go modules separately so this doesn't re-run on every change
 WORKDIR /src
+# copy in our certs
+COPY certs/localhost.cert.pem . 
+COPY certs/localhost.key.pem . 
+COPY certs/ca-chain.cert.pem .
```
Comment (taylorsilva) [inline Dockerfile:27]: Can we generate these inside this Dockerfile, similar to the above lines? That way we avoid having to commit any certs to the repo or the certs expiring.

**Review by taylorsilva — CHANGES_REQUESTED** [review]
Left a bunch of comments! Let me know your thoughts :)

File: certs/ca-chain.cert.pem [hunk 3]
```diff
@@ -0,0 +1,65 @@
+-----BEGIN CERTIFICATE-----
```
[the remark this answers is not in the document]
Comment (taylorsilva) [reply certs/ca-chain.cert.pem:1]: Going to partially back out this suggestion but still suggest things moves under the `hack/` directory. If you checkout master now you'll see some certs are stored in various folders inside `hack/`

File: docker-compose.yml [hunk 4]
```diff
@@ -16,17 +16,17 @@ services:
     image: concourse/concourse:local
     volumes:
     - .:/src
-    command: web
+    command: web --external-url https://localhost:443  --tls-bind-port 8443  --tls-cert /src/certs/localhost.cert.pem  --tls-key  /src/certs/localhost.key.pem --tls-ca-cert /src/certs/ca-chain.cert.pem
```
[the remark this answers is not in the document]
Comment (nickhyoti) [reply docker-compose.yml:19]: Hi 

I've added this to our branch, with a reverse proxy as well.

File: release-notes/latest.md [hunk 5]
```diff
@@ -33,3 +33,5 @@
 
   A new metric called `tasks_wait_duration_bucket` is also added to express as quantiles the average time spent by tasks awaiting execution. PR: #5981
   ![Example graph for the task wait time histograms.](https://user-images.githubusercontent.com/40891147/89990749-189d2600-dc83-11ea-8fde-ae579fdb0a0a.png)
+  #### <sub><sup><a name="5981" href="#5981">:link:</a></sup></sub> feature
+  * Added optional support for mTLS to mandate use of certificate from reverse proxy to connect to Concourse.  RFC raised for this earlier this year https://github.com/concourse/rfcs/pull/73
```
[the remark this answers is not in the document]
Comment (nickhyoti) [reply release-notes/latest.md:37]: Hi,

Can you clarify - should the release_notes directory be deleted i.e. git rm -r release_notes?

File: yarn.lock [hunk 6]
```diff
@@ -9,24 +9,28 @@
   dependencies:
     "@babel/highlight" "^7.10.4"
 
-"@babel/compat-data@^7.12.0":
-  version "7.12.0"
-  resolved "https://registry.yarnpkg.com/@babel/compat-data/-/compat-data-7.12.0.tgz#443aea07a5aeba7942cb067de6b8272f2ab36b9e"
-  integrity sha512-jAbCtMANC9ptXxbSVXIqV/3H0bkh7iyyv6JS5lu10av45bcc2QmDNJXkASZCFwbBt75Q0AEq/BB+bNa3x1QgYQ==
+"@babel/compat-data@^7.10.4", "@babel/compat-data@^7.11.0":
```
[the remark this answers is not in the document]
Comment (nickhyoti) [reply yarn.lock:12]: Hi,

I did a rebase and committed to my branch

File: testflight/suite_test.go (hunk 1 above)
Comment (nickhyoti) [reply testflight/suite_test.go]: removed

File: certs/ca-chain.cert.pem (hunk 3 above)
[the remark this answers is not in the document]
Comment (nickhyoti) [reply certs/ca-chain.cert.pem:1]: Hi,

Can you clarify on the preferred approach here?  Are you OK for me to create certificates (with a long expiration?) and commit those?

Thanks

File: release-notes/latest.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (taylorsilva) [reply release-notes/latest.md:37]: You can ignore that entire folder. If you rebase on master it should disappear for your local branch. Might be easier for you to remove or amend the commit that modifies the file. I'm going to update your PR description to add the Release Notes section

[the remark this answers is not in the document]
Comment (taylorsilva) [reply release-notes/latest.md:37]: Actually I saw it there already, so I copied over your release note text for you :)

File: certs/ca-chain.cert.pem (hunk 3 above)
[the remark this answers is not in the document]
Comment (taylorsilva) [reply certs/ca-chain.cert.pem:1]: I'm fine with your approach of adding certs to the repo for easier testing :) no need to change this at all now. Just move the location of the certs. You should find a bunch under the `hack/keys` directory now.

```
$ ls -l hack/keys
total 56
-rw-r--r--  1 tasilva  staff  213 29 Jan 10:19 authorized_worker_keys
-rw-r--r--  1 tasilva  staff  879 29 Jan 10:19 session_signing_key
-rwxr-xr-x  1 tasilva  staff  250 29 Jan 10:19 setup
-rw-r--r--  1 tasilva  staff  883 29 Jan 10:19 tsa_host_key
-rw-r--r--  1 tasilva  staff  213 29 Jan 10:19 tsa_host_key.pub
-rw-r--r--  1 tasilva  staff  879 29 Jan 10:19 worker_key
-rw-r--r--  1 tasilva  staff  213 29 Jan 10:19 worker_key.pub
```

File: release-notes/latest.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (taylorsilva) [reply release-notes/latest.md:37]: Honestly though, this isn't a blocker. Everything under that folder is ignored anyways :)

File: docker-compose.yml [hunk 7]
```diff
@@ -54,3 +54,4 @@ services:
       CONCOURSE_BIND_IP: 0.0.0.0
       CONCOURSE_BAGGAGECLAIM_BIND_IP: 0.0.0.0
       CONCOURSE_GARDEN_DNS_PROXY_ENABLE: "true"
+
```
Comment (taylorsilva) [inline docker-compose.yml:57]: nit: remove this blank line please

File: hack/overrides/mutual-tls.yml [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+      CONCOURSE_MAIN_TEAM_LOCAL_USER: test
+      CONCOURSE_CLUSTER_NAME: dev
+      CONCOURSE_ENABLE_PIPELINE_INSTANCES: "true"
+      CONCOURSE_ENABLE_ACROSS_STEP: "true"
+      CONCOURSE_TLS_BIND_PORT: 8443
+      CONCOURSE_TLS_CERT: /src/certs/web.cert.pem 
+      CONCOURSE_TLS_CA_CERT: /src/certs/ca-chain.cert.pem
+      CONCOURSE_TLS_KEY: /src/certs/web.key.pem
+
+  nginx_mtls_rp:
```
Comment (taylorsilva) [inline hack/overrides/mutual-tls.yml:30]: nit:
```suggestion
  nginx:
```

File: hack/overrides/mutual-tls.yml [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+services:
+  web:
+    build: .
+    image: concourse/concourse:local
+    volumes:
+    - .:/src
+    command: web
+    depends_on: [db]
+    #ports:
+    #- 8443:8443
```
Comment (taylorsilva) [inline hack/overrides/mutual-tls.yml:12]: nit
```suggestion
```

File: Dockerfile [hunk 10]
```diff
@@ -22,6 +22,10 @@ ENV CONCOURSE_TSA_WORKER_PRIVATE_KEY  /concourse-keys/worker_key
 
 # download go modules separately so this doesn't re-run on every change
 WORKDIR /src
+# copy in our certs
+#COPY certs/web.cert.pem . 
+#COPY certs/web.key.pem . 
+#COPY certs/ca-chain.cert.pem .
```
Comment (taylorsilva) [inline Dockerfile:28]: nit:
```suggestion
```

**Review by taylorsilva — CHANGES_REQUESTED** [review]
This is looking good! Just a few nits (nit-picks) as some things seem like they're in a "hacky" state. It would be nice to clean them up!

I wanted to try out the changes,so I ran:
```
docker-compose -f docker-compose.yml -f ./hack/overrides/mutual-tls.yml up -d --build
```
But both the web and nginx container failed to come up :(
```diff
docker ps -a
CONTAINER ID   IMAGE                       COMMAND                  CREATED         STATUS                     PORTS                                            NAMES
-957ac444b776   nginx:latest                "/docker-entrypoint.…"   8 minutes ago   Exited (1) 6 minutes ago                                                    concourse_nginx_mtls_rp_1
e984bda7e4fc   concourse/concourse:local   "dumb-init concourse…"   8 minutes ago   Up 8 minutes               0.0.0.0:7777->7777/tcp, 0.0.0.0:7788->7788/tcp   concourse_worker_1
-2c99bd13c46d   concourse/concourse:local   "dumb-init concourse…"   8 minutes ago   Exited (1) 6 minutes ago                                                    concourse_web_1
e22d2baaeb6c   postgres:latest             "docker-entrypoint.s…"   8 minutes ago   Up 8 minutes               0.0.0.0:6543->5432/tcp                           concourse_db_1
```

File: docker-compose.yml (hunk 7 above)
Comment (nickhyoti) [reply docker-compose.yml]: deleted

File: hack/overrides/mutual-tls.yml [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+      CONCOURSE_TLS_CERT: /src/certs/web.cert.pem 
+      CONCOURSE_TLS_CA_CERT: /src/certs/ca-chain.cert.pem
+      CONCOURSE_TLS_KEY: /src/certs/web.key.pem
+
+  nginx_mtls_rp:
+    image: nginx:latest
+    depends_on: [web]
+    volumes:
+    - ./hack/nginx/conf.d:/etc/nginx/conf.d
+    - ./hack/mtls_certs:/etc/nginx/mtls_certs
```
[the remark this answers is not in the document]
Comment (nickhyoti) [reply hack/overrides/mutual-tls.yml:33]: This should now be running/runnable.

I cloned down into a clean environment and it started for me

File: hack/overrides/mutual-tls.yml (hunk 8 above)
Comment (nickhyoti) [reply hack/overrides/mutual-tls.yml]: changed

File: hack/overrides/mutual-tls.yml (hunk 9 above)
Comment (nickhyoti) [reply hack/overrides/mutual-tls.yml]: deleted

File: Dockerfile (hunk 10 above)
Comment (nickhyoti) [reply Dockerfile]: deleted

File: hack/overrides/mutual-tls.yml (hunk 11 above)
[the remark this answers is not in the document]
Comment (taylorsilva) [reply hack/overrides/mutual-tls.yml:33]: It works now! yay! I can play with it 😁

**Review by taylorsilva — CHANGES_REQUESTED** [review]
Almost there! Some cleanup around the docker-compose override file and adjusting the cert expiration to some time in the far future.

File: docker-compose.yml [hunk 12]
```diff
@@ -20,6 +20,7 @@ services:
     depends_on: [db]
     ports:
     - 8080:8080
+    - 9102:9102
```
[the remark this answers is not in the document]
Comment (nickhyoti) [reply docker-compose.yml]: Removed

File: hack/overrides/mutual-tls.yml [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+      CONCOURSE_POSTGRES_HOST: db
+      CONCOURSE_POSTGRES_USER: dev
+      CONCOURSE_POSTGRES_PASSWORD: dev
+      CONCOURSE_POSTGRES_DATABASE: concourse
+      CONCOURSE_EXTERNAL_URL: https://localhost:443
+      CONCOURSE_ADD_LOCAL_USER: test:test,guest:guest
+      CONCOURSE_MAIN_TEAM_LOCAL_USER: test
+      CONCOURSE_CLUSTER_NAME: dev
+      CONCOURSE_ENABLE_PIPELINE_INSTANCES: "true"
+      CONCOURSE_ENABLE_ACROSS_STEP: "true"
```
[the remark this answers is not in the document]
Comment (nickhyoti) [reply hack/overrides/mutual-tls.yml]: removed

File: hack/mtls_certs/backend/web.key.pem [hunk 14]
```diff
@@ -0,0 +1,27 @@
+-----BEGIN RSA PRIVATE KEY-----
```
[the remark this answers is not in the document]
Comment (nickhyoti) [reply hack/mtls_certs/backend/web.key.pem:1]: Extended to 2121

File: hack/overrides/mutual-tls.yml [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+    volumes:
+    - .:/src
+    command: web
+    depends_on: [db]
+    environment:
+      CONCOURSE_LOG_LEVEL: debug
+      CONCOURSE_POSTGRES_HOST: db
+      CONCOURSE_POSTGRES_USER: dev
+      CONCOURSE_POSTGRES_PASSWORD: dev
+      CONCOURSE_POSTGRES_DATABASE: concourse
```
[the remark this answers is not in the document]
Comment (nickhyoti) [reply hack/overrides/mutual-tls.yml]: Removed

**Comment by nickhyoti** [comment]
I see that the DCO signoff needs to be actioned.

I tried this myself but got the following
```
$ git rebase HEAD~29 --signoff
First, rewinding head to replay your work on top of it...
Applying: web: structure: use pipeline's database id as identifier
Using index info to reconstruct a base tree...
M	web/elm/src/Build/Build.elm
M	web/elm/src/Concourse.elm
M	web/elm/src/Dashboard/Dashboard.elm
M	web/elm/src/Dashboard/Pipeline.elm
M	web/elm/src/Message/Effects.elm
M	web/elm/src/Pipeline/PinMenu/PinMenu.elm
M	web/elm/src/Pipeline/Pipeline.elm
M	web/elm/src/Resource/Resource.elm
M	web/elm/src/SideBar/Pipeline.elm
M	web/elm/src/SideBar/SideBar.elm
M	web/elm/src/SideBar/Team.elm
M	web/elm/src/SubPage/SubPage.elm
Falling back to patching base and 3-way merge...
Auto-merging web/elm/src/SubPage/SubPage.elm
Auto-merging web/elm/src/SideBar/Team.elm
Auto-merging web/elm/src/SideBar/SideBar.elm
CONFLICT (content): Merge conflict in web/elm/src/SideBar/SideBar.elm
Auto-merging web/elm/src/SideBar/Pipeline.elm
Auto-merging web/elm/src/Resource/Resource.elm
CONFLICT (content): Merge conflict in web/elm/src/Resource/Resource.elm
Auto-merging web/elm/src/Pipeline/Pipeline.elm
Auto-merging web/elm/src/Pipeline/PinMenu/PinMenu.elm
Auto-merging web/elm/src/Message/Effects.elm
Auto-merging web/elm/src/Dashboard/Pipeline.elm
Auto-merging web/elm/src/Dashboard/Dashboard.elm
CONFLICT (content): Merge conflict in web/elm/src/Dashboard/Dashboard.elm
Auto-merging web/elm/src/Concourse.elm
CONFLICT (content): Merge conflict in web/elm/src/Concourse.elm
Auto-merging web/elm/src/Build/Build.elm
error: Failed to merge in the changes.
Patch failed at 0001 web: structure: use pipeline's database id as identifier
Use 'git am --show-current-patch' to see the failed patch

Resolve all conflicts manually, mark them as resolved with
"git add/rm <conflicted_files>", then run "git rebase --continue".
You can instead skip this commit: run "git rebase --skip".
To abort and get back to the state before "git rebase", run "git rebase --abort".
```

Whats the best course for this?

**Comment by taylorsilva** [comment]
Must be the merge commits. I set DCO to pass because I do see that you signed off on them

**Comment by muntac** [comment]
Rerunning `testflight` with the latest dev image (with runc rc 92). Sorry for dropping the ball on this. Thanks a ton for picking this up @taylorsilva!

**Comment by muntac** [comment]
hm worker glitch this time. rerunning - https://ci.concourse-ci.org/teams/contributor/pipelines/prs/jobs/testflight/builds/126.1

**Comment by muntac** [comment]
Aaand looks like the whole PRs pipeline has disappeared  : / - https://ci.concourse-ci.org/teams/contributor/pipelines/prs/

**Comment by taylorsilva** [comment]
@muntac seems like the UI changed. The URL to the pipeline is now this search term: https://ci.concourse-ci.org/?search=team%3A%22contributor%22%20group%3A%22prs%22

**Comment by aoldershaw** [comment]
@taylorsilva 

> seems like the UI changed. The URL to the pipeline is now this search term

@muntac's link is valid, but github stripped didn't include the last `"` in the hyperlink. The link is `https://ci.concourse-ci.org/teams/contributor/pipelines/prs/?vars.branch="master"`, but if you clicked the hyperlink, it brought you to `https://ci.concourse-ci.org/teams/contributor/pipelines/prs/?vars.branch="master`

**Comment by taylorsilva** [comment]
@aoldershaw I only saw `https://ci.concourse-ci.org/teams/contributor/pipelines/prs/`, the URL he crossed out. My comment wasn't to his edit :)

**Review by taylorsilva — APPROVED** [review]
LGTM.
Thanks for all your work! :)

**Comment by nickhyoti** [comment]
Thanks for your help on getting this through

**Pull request merged**